src/commissioner/Events.h

```diff
@@ -0,0 +1,316 @@
+/*
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <controller/CommissioneeDeviceProxy.h>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <protocols/secure_channel/CASESession.h>
+#include <protocols/secure_channel/PASESession.h>
+#include <protocols/secure_channel/RendezvousParameters.h>
+#include <setup_payload/ManualSetupPayloadParser.h>
+#include <setup_payload/QRCodeSetupPayloadParser.h>
+#include <setup_payload/SetupPayload.h>
+
+namespace chip {
+namespace Commissioner {
+namespace Events {
+
+constexpr size_t kMaxAttestationElementsLength = 900; // RESP_MAX
+constexpr size_t kMaxNocsrElementsLength       = 900; // RESP_MAX
+constexpr size_t kMaxNetworkIdLength           = 32;
+
+struct SharedBuffer
+{
+    const ByteSpan Get() const { return ByteSpan(mData.get(), mLen); }
+    MutableByteSpan GetMutable() { return MutableByteSpan(mData.get(), mCapacity); };
+    CHIP_ERROR Set(const ByteSpan buffer)
+    {
+        if (mCapacity < buffer.size())
+        {
+            ReturnErrorOnFailure(Allocate(buffer.size()));
+        }
+        memmove(mData.get(), buffer.data(), buffer.size());
+        mLen = buffer.size();
+        return CHIP_NO_ERROR;
+    }
+    void Clear()
+    {
+        mData     = std::shared_ptr<uint8_t>(nullptr);
+        mCapacity = 0;
+        mLen      = 0;
+    }
+    CHIP_ERROR Allocate(size_t size)
+    {
+        mData = std::shared_ptr<uint8_t>(static_cast<uint8_t *>(chip::Platform::MemoryAlloc(size)), chip::Platform::MemoryFree);
```

Comment:
Given this class is subclassed by `CapacityBoundSharedBuffer`, which is only used by fixed-sized things, can `MemoryAlloc` be avoided?

Reply:
`MemoryAlloc` is intentionally used here to reduce the high-water size for all events and states, as all pay the penalty for their largest sized member types in the variants.

Specifically, these types seemed appropriate for dynamic allocation:

```
    CapacityBoundSharedBuffer<Credentials::kMaxDERCertLength> mPai;
    CapacityBoundSharedBuffer<Credentials::kMaxDERCertLength> mDac;
    CapacityBoundSharedBuffer<Credentials::kMaxCHIPCertLength> mRcac;
    CapacityBoundSharedBuffer<Credentials::kMaxCHIPCertLength> mIcac;
    CapacityBoundSharedBuffer<Credentials::kMaxCHIPCertLength> mNoc;
```

In my defense, CommissioneeDeviceProxy also currently dynamically allocates the PAI and DAC:

https://github.com/project-chip/connectedhomeip/blob/a0e8e3a94b4b27aaeb2e0af42b9ecdebd84be269/src/controller/CommissioneeDeviceProxy.cpp#L255
https://github.com/project-chip/connectedhomeip/blob/a0e8e3a94b4b27aaeb2e0af42b9ecdebd84be269/src/controller/CommissioneeDeviceProxy.cpp#L219

Regardless of the use of MemoryAlloc or not, the design is thrifty in that the state machine only maintains items resident in memory when they are needed.  The CommissioneeDeviceProxy has the NOC and ICA resident in memory for its entire lifetime:

```
 uint8_t mNOCCertBuffer[Credentials::kMaxCHIPCertLength];
 uint8_t mICACertBuffer[Credentials::kMaxCHIPCertLength];
```

Given the 'high-water' penalty, I feel it would be unfortunate to significantly enlarge the Event and State variant types in all targets.  One thing I've thought about though is having a compile-time switch to choose between static and dynamic allocation for the entire state machine.

I will explore that.  If it is feasible, would that be an acceptable solution?